cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -738,6 +761,109 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
 
 }  // namespace
 
+Status RegisterFileSystemFactory(std::vector<std::string> schemes,
+                                 FileSystem::Factory factory) {
+  auto& [mutex, scheme_to_factory] = FileSystemFactoryRegistry();
+  std::unique_lock lock{mutex};
+
+  for (auto&& scheme : schemes) {
+    if (auto& ref = scheme_to_factory[scheme]) {
+      return Status::KeyError(
+          "Tried to add a factory for ", scheme,
+          ":// uris, but a factory was already registered for that scheme");
+    } else {
+      ref = factory;
+    }
+  }
+  return Status::OK();
+}
+
+// XXX alternative approach:
```

Comment:
I like this approach.
Because hiding `dlopen()`/`LoadLibrary()` details from users is easy to use.
(Users need to call `dlopen()`/`LoadLibrary()` explicitly with the `Registrar` approach, right?)